---
title: "Praktinė užduotis Nr. 2"
subtitle: "Klasterizavimas"
author: "Jaroslav Rutkovskij, Danielė Stasiūnaitė, Karolis Augustauskas (JDK)"
date: "4/24/2022"
output: pdf_document
---
```{r Library preparations, include=FALSE}
library(Signac)
library(dplyr)
library(heatmaply)
library(reshape2)
library(ggplot2)
library(data.table)
```

```{r Data_access, include=FALSE}
beta <- readRDS("beta_values.rds")            # Modification value matrix
pData_col <- readRDS("df.rds")      # Sample (column) matrix
annot_row <- readRDS("name.rds") # Position (row) matrix
```

# 1. Mėginių hierarchinis klasterizavimas
```{r}
```

# 2. Mėginių atvaizdavimas "heatmap" pavidalu
```{r Task_2_1, include=FALSE}
# Determining unique 'chr' column values (chromosome names) from the matrix
# that describes positions (rows):
chr_names <- unique(annot_row$chr)     
genes_only <- c()

# Loop which checks whether there are common genes between chromosome pairs:
for (i in 1:length(chr_names)) {
  for (y in 1:length(chr_names)) {
    if (i == y) { next }
    else {
      names1 <- annot_row[annot_row$chr == chr_names[i],
                          'UCSC_RefGene_Name'] %>% unique()
      names2 <- annot_row[annot_row$chr == chr_names[y],
                          'UCSC_RefGene_Name'] %>% unique()
      genes <- intersect(names1, names2)
      if (genes[1] == '' & length(genes) == 1) { next }
      else {
        # print(paste0(i, " ir ", y, ": ", genes))
        genes_only <- c(genes_only, genes)
      }
    }
  }
}
```
```{r Task_2_2, include=FALSE}
# Removing empty elements ('') from the gene list:
genes_only <- genes_only[! genes_only %in% c('')]

# Creating gene expression frequency table:
table(genes_only)

gene_list <- c(unique(genes_only))
gene_cnts <- data.frame(chr = chr_names)  # gene_conts --> gene count dataframe
gene_cnts
for (g in 1:length(gene_list)) {
  gene <- c()
  for (j in 1:length(chr_names)) {
    annot_chr <- annot_row[annot_row$chr == chr_names[j], ]
    annot_gene <- annot_chr[annot_chr$UCSC_RefGene_Name == gene_list[g], ] %>% 
                              data.matrix()
    #if (length(annot_gene) == 0) {next}
    # print(paste0(gene_list[g], ", ", chr_names[j], " = ", length(annot_gene)))
    gene <- c(gene, length(annot_gene))
  }
  gene_cnts[ , ncol(gene_cnts) + 1] <- gene
  colnames(gene_cnts)[ncol(gene_cnts)] <- gene_list[g]
}
```
```{r Task_2_3, include=FALSE}
gene_cmt <- data.matrix(gene_cnts)     # gene_cmt --> gene count matrix
gene_cmt <- gene_cmt[ , 2:length(colnames(gene_cmt))]

# fltrd_gene_cnts --> filtered gene count matrix (rows that contained only
# only zero values were removed. It means that gene expression in
# particular chromosome was not identified):
fltrd_gene_cnts <- gene_cmt
rownames(fltrd_gene_cnts) <- c(chr_names)
fltrd_gene_cnts <- fltrd_gene_cnts[rowSums(fltrd_gene_cnts[]) > 0, ]
gene_cnts_melt <- melt(fltrd_gene_cnts)

plot1 <- ggplot(gene_cnts_melt, aes(x = Var2, y = value))+
            geom_bar(stat = 'identity', fill = "#123e68")+
            ylim(0, 1700) +
            facet_wrap(~Var1) +
            labs(title = "Genų ekspresija chromosomose", x = "Genas",
                 y = "Ekspresija") +
            geom_text(aes(label = value), vjust = -0.5, color = "black",
                      size = 3)+
            theme(axis.text.x = element_text(angle = 45, vjust = 0.5),
                  plot.title = element_text(hjust = 0.5, face = "bold"))
plot1
```
Sukūrus aukščiau pavaizduotas stulpelines diagramas buvo nustatyta, kokie genai
buvo dažniausi bei kokioje chromosomoje jų buvo daugiausiai.

```{r Task_2_4, include=FALSE}

```

Nuskaitytiems duomenims pritaikyta standartinė ***heatmap()*** funkcija,
iliustruojanti **beta** matricos reikšmes ir pagal spalvų intensyvumą bei
atspalvį, leidžianti daryti išvadas apie duomenų grupes - klasterius.\
Kadangi **beta** matrica turi ***`r dim(beta)[1]`*** eilutes bei
***`r dim(beta)[2]`*** stulpelius, pavaizduoti visas matricos reikšmes,
naudojantis ***heatmap()*** funkcija yra negalima.





# 3. Tikro ir nuspėto amžiaus bei senėjimo "pagreitėjimo" nustatymas
```{r}
#library(BiocManager)
#install(c("tidyverse", "impute", "Rcpp"))
#install.packages("devtools", dependencies = TRUE)

#BiocManager::install("methylclock")
#BiocManager::install("methylclockData")
library(methylclock)
library(methylclockData)

library(Biobase)
library(tibble)
library(impute)
library(ggplot2)
library(ggpmisc)
library(GEOquery)

df1 <- data.matrix(beta)

cpgs.missing <- checkClocks(df1)
cpgs.missing.GA <- checkClocksGA(df1)

df2<-matrix(c(rownames(df1), df1,df1),dim(df1))

df2
dim(df2)

df3 <-data.frame(df2)

dim(df3)
DNAmAge(df3,clocks="Hannum")

```
